design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+
+|      | Deployment                  | Elaboration                                                  | Pros                                                         | Cons                                                         |
+| ---- | --------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Run as Job                  | Run as a K8s Job that deploys a Pod that terminates after task completion | - Minimal resources usage<br />- Minimal code changes<br />- Explicit user experience (user expects Job to run to completion quickly) | - "Job" is not an existing application identity granularities. Adding it can take low priority since customers can run the Job using Namespace / Service Account granularities<br /><br />- Deployment will need to be changed (from Job → Pod) for Milestone 2 (support secrets rotation) |
+| 2    | Run as non-terminating  Pod | Run as a Pod that does not terminate automatically after task completion | - Simple to below<br />                                      | - Wastes resources because we aren't updating K8s secrets and up all the time<br /><br />- For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes. (e.g. secrets provider as init-container, conjur authn k8s secrets, any of our clients or Conjur REST API) |
+
+**Decision:** Run as a Job
+
+We decided to run as a Job instead of a Pod because for Milestone 1 we are concerned with delivering Conjur values only once. At this time we are not supporting updating K8s Secrets with Conjur secret across multiple runs so deploying a Pod that will run continuously in the environment of a customer is a waste of their resources. Especially because the process of retrieving secrets will only happen once, at the inital spin up of the Secrets Provider.
+
+#### Flow
+
+The flow for Milestone 1 is similar to the user experience from Phase 1. For each step, we will go into more detail below.
+
+1. Configure K8s Secrets with ***conjur-map\*** metadata defined 
+
+   - *Optional:* add label to K8s Secret ***(Milestone 1)***
+
+2. Create Service Account for Secrets Provider 
+
+3. Create Role / ClusterRole for the Service Accounts with **get**/**update** privileges
+
+4. Create RoleBinding / ClusterRoleBinding and bind the Service Accounts to the role from previous step
+
+5. Create/Deploy Secrets Providers as Jobs in deployment manifest(s) ***(Milestone 1)***
+   - If label filtering is defined, add "K8S_SECRET_LABEL" in the Secrets Provider manifest(s) ***(Milestone 1)***
```

Comment:
do we need this label to be configurable?

Reply:
I get to it later in the doc that the key for the label can be configurable